es/index.js

```diff
@@ -59,6 +59,21 @@ export function isNetworkOrIdempotentRequestError(error) {
   return isNetworkError(error) || isIdempotentRequestError(error);
 }
 
+/**
+ * @return {number} - delay in milliseconds, always 1
+ */
+export function noDelay() {
+  return 1;
+}
+
+/**
+ * @param  {number} [retryNumber=0]
+ * @return {number} - delay in milliseconds
+ */
+export function exponentialDelay(retryNumber = 0) {
+  return (Math.pow(2, retryNumber) * 1000) + Math.floor(Math.random() * 1000);
```

Comment:
This starts with a random delay between 0 and 1000, which I think it may be too much. I think the first retry should have no delay and then start with a low value (e.g.: ~100ms)

Reply:
Hi,

This function returns a value between min and max for the given retry count:

| Retry count | Min | Max |
| -- | -- | -- |
| 1 | 2000 | 3000 |
| 2 | 4000 | 5000 |
| 3 | 8000 | 9000 |
| 4 | 16000 | 17000 |
| 5 | 32000 | 33000 |
| 6 | 64000 | 65000 |
| 7 | 128000 | 129000 |
| 8 | 256000 | 257000 |
| 9 | 512000 | 513000 |
| 10 | 1024000 | 1025000 |

This does not look correct to me and should be:

| Retry count | Min | Max |
| -- | -- | -- |
| 1 | 1000 | 2000 |
| 2 | 2000 | 3000 |
| 3 | 4000 | 5000 |
| 4 | 8000 | 9000 |
| 5 | 16000 | 17000 |
| 6 | 32000 | 33000 |
| 7 | 64000 | 65000 |
| 8 | 128000 | 129000 |
| 9 | 256000 | 257000 |
| 10 | 512000 | 513000 |

Which then fits with this [exponential backoff](https://developers.google.com/analytics/devguides/reporting/core/v3/errors#backoff) explanation.

I take your point on the initial retry, but I would say that is a different function.

Reply:
The exponential backoff is a technique, not a specific function or algorithm. The one you link is one possible implementation for Google Analytics, which is meant to be used in the client and no special data is expected from it, so I don't think it's the best implementation to choose. As I said, I'd modify the implementation to start with 100ms and increase from there. Something like:

```javascript
export function exponentialDelay(retryNumber = 0) {
  const delay = Math.pow(2, retryNumber) * 100;
  const randomSum = delay * 0.2 * Math.random(); // 0-20% of the delay
  return delay + randomSum;
}
```

This way we'll have:

| Retry count | Min | Max |
| -- | -- | -- |
| 0 | 100 | 120 |
| 1 | 200 | 240 |
| 2 | 400 | 480 |
| 3 | 800 | 960 |
| 4 | 1600 | 1920 |
| 5 | 1600 | 3840 |

Similar to other popular implementations (such as [this one from Amazon](https://docs.aws.amazon.com/general/latest/gr/api-retries.html)).

Reply:
fair enough